public/app/features/alerting/unified/api/preview.ts

```diff
@@ -0,0 +1,50 @@
+import { dataFrameFromJSON, DataFrameJSON, getDefaultTimeRange, LoadingState } from '@grafana/data';
+import { getBackendSrv } from '@grafana/runtime';
+import {
+  CloudPreviewRuleRequest,
+  GrafanaPreviewRuleRequest,
+  isCloudPreviewRequest,
+  isGrafanaPreviewRequest,
+  PreviewRuleRequest,
+  PreviewRuleResponse,
+} from '../types/preview';
+import { RuleFormType } from '../types/rule-form';
+
+export async function previewAlertRule(request: PreviewRuleRequest): Promise<PreviewRuleResponse> {
+  if (isCloudPreviewRequest(request)) {
+    return previewCloudAlertRule(request);
+  }
+
+  if (isGrafanaPreviewRequest(request)) {
+    return previewGrafanaAlertRule(request);
+  }
+
+  throw new Error('unsupported preview rule request');
+}
+
+type GrafanaPreviewRuleResponse = {
+  instances: DataFrameJSON[];
+};
+
+async function previewGrafanaAlertRule(request: GrafanaPreviewRuleRequest): Promise<PreviewRuleResponse> {
```

Comment:
I have two different methods here because we are getting the response in different formats.

Reply:
Why are we using Promises instead of streams? `toPromise` is a deprecated function that will be removed in RxJs 7.0.

Reply:
In this case, because it is less code and a super simple scenario.

Reply:
looks like the upgrade is straightforward `await fooOsbervable.toPromise()` -> `await lastValueFrom(fooObservable)`

Reply:
I don't agree but I'm ok with your answers @mckn @domasx2

Reply:
@hugohaggmark what do you think about my updates? turned out to be pretty nice with loading state being displayed if request takes longer then 200ms.

Reply:
@mckn beautiful 
![](https://media3.giphy.com/media/3ofT5ySFXZ01oJ7aoM/giphy.gif)